web/src/server/api/routers/utilities.ts

```diff
@@ -1,19 +1,76 @@
 import { createTRPCRouter, protectedProcedure } from "@/src/server/api/trpc";
 import { z } from "zod";
+import { promises as dns } from "dns";
+import { Address4, Address6 } from "ip-address";
 
-const isValidURL = (urlString: string): boolean => {
+const IP_4_LOOPBACK_SUBNET = "127.0.0.0/8";
+const IP_4_LINK_LOCAL_SUBNET = "169.254.0.0/16";
+const IP_4_PRIVATE_A_SUBNET = "10.0.0.0/8";
+const IP_4_PRIVATE_B_SUBNET = "172.16.0.0/12";
+const IP_4_PRIVATE_C_SUBNET = "192.168.0.0/16";
+
+const isPrivateIp = (hostname: string): boolean => {
+  try {
+    if (Address6.isValid(hostname)) {
+      const address = new Address6(hostname);
+      return address.isLinkLocal() || address.isLoopback();
+    } else if (Address4.isValid(hostname)) {
+      const address = new Address4(hostname);
+      return [
+        IP_4_LOOPBACK_SUBNET,
+        IP_4_LINK_LOCAL_SUBNET,
+        IP_4_PRIVATE_A_SUBNET,
+        IP_4_PRIVATE_B_SUBNET,
+        IP_4_PRIVATE_C_SUBNET,
+      ].some((subnet) => address.isInSubnet(new Address4(subnet)));
+    } else {
+      console.error("Invalid IP address:", hostname);
+      return false;
+    }
+  } catch (error) {
+    console.error("IP parsing error:", error);
+    return false;
+  }
+};
+
+const resolveHostname = async (hostname: string): Promise<string[]> => {
+  try {
+    const addresses = await dns.resolve4(hostname);
```

Comment:
The built-in dns module in Node.js does not support specifying a DNS server directly ([docs](https://nodejs.org/api/dns.html#dnsresolve4hostname-options-callback)). We could use an external package such as `dns-packet` (combined w/ `dgram` built into node) to support this. Currently it would default to the users networks DHCP server.